### PR TITLE
add profile in offline benchmark & update doc

### PR DESCRIPTION
**Pull request by bjmsong**
## Modifications

- add profile in offline benchmark
- update doc

## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/bench_offline_throughput.py [hunk 1]
```diff
@@ -156,6 +160,20 @@ def add_cli_args(parser: argparse.ArgumentParser):
             action="store_true",
             help="Do not exit the program. This is useful for nsys profile with --duration and --delay.",
         )
+        parser.add_argument(
+            "--profile",
+            action="store_true",
+            help="profile the generation process of a single batch",
```
Comment (merrymercy) [inline python/sglang/bench_offline_throughput.py]: It is not a single batch. Should we use the doc string here https://github.com/sgl-project/sglang/blob/2369e88209afca1e431d355bc86f7deae08e4b7c/python/sglang/bench_serving.py#L1402-L1403?

File: python/sglang/bench_offline_throughput.py [hunk 2]
```diff
@@ -194,7 +214,23 @@ def throughput_test_once(
     ]
 
     st = time.perf_counter()
-    gen_out = backend.generate(prompt=prompt, sampling_params=sampling_params)
+    if profile:
+        if not profile_dir:
+            profile_dir = (
+                Path(".") / "vllm_benchmark_result" / f"latency_result_{time.time()}"
```
Comment (merrymercy) [inline python/sglang/bench_offline_throughput.py]: vllm -> sglang?

File: python/sglang/bench_offline_throughput.py [hunk 3]
```diff
@@ -268,6 +304,8 @@ def throughput_test(
             reqs=warmup_requests,
             ignore_eos=not bench_args.disable_ignore_eos,
             extra_request_body=extra_request_body,
+            profile=bench_args.profile,
```
Comment (merrymercy) [inline python/sglang/bench_offline_throughput.py]: do not need to profile the warmup?

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/bench_offline_throughput.py (hunk 3 above)
Comment (bjmsong) [reply python/sglang/bench_offline_throughput.py]: yes, indeed

File: python/sglang/bench_offline_throughput.py [hunk 4]
```diff
@@ -194,7 +214,23 @@ def throughput_test_once(
...
 
     st = time.perf_counter()
-    gen_out = backend.generate(prompt=prompt, sampling_params=sampling_params)
+    if profile:
+        if not profile_dir:
+            profile_dir = (
+                Path(".") / "sglang_benchmark_result" / f"latency_result_{time.time()}"
+            )
+        print(f"Profiling (results will be saved to '{profile_dir}')...")
+        with torch.profiler.profile(
```
Comment (merrymercy) [inline python/sglang/bench_offline_throughput.py]: backend.generate will call into another process. Can this torch profiler context work across processes?

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/bench_offline_throughput.py (hunk 4 above)
Comment (bjmsong) [reply python/sglang/bench_offline_throughput.py]: To My Knowledge,  torch profiler can't work across processes. So the profiler was changed to scheduler process.

File: python/sglang/srt/server.py [hunk 5]
```diff
@@ -168,6 +168,16 @@ async def flush_cache():
     )
 
 
+def start_profile_sync():
```
Comment (merrymercy) [inline python/sglang/srt/server.py]: remove `_sync` in the name. If it conflicts with the name below, change the names of the fastapi functions below.

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/server.py (hunk 5 above)
Comment (bjmsong) [reply python/sglang/srt/server.py]: done

**Pull request merged**